### DISCUSSION TITLE
Handle long running requests

### DISCUSSION THREAD

**Discussion by ghost** [rwf2/Rocket#2368] [Questions]
The service I'm building processes data on the server and takes long time (3mins) to finish and return a result.

What's best way to handle such cases?
Maybe return, do processing, and call some callback?
Is it possible to do such thing with Rocket?

**Comment by wezm** [comment]
Rocket 0.5 is built on async Rust and will handle this quite gracefully. You probably want to spawn a tokio task for the long running process and then await that. tokio will run that task on its pool of threads and the rest of the Rocket process will continue handling requests. Something like this:

```rust
use rocket::tokio;

#[get("/something")]
async fn your_handler() -> YourResponseType {
    let res = tokio::task::spawn_blocking(move || {
        // long running stuff in here
    })
    .await;

    //  Generate response using result of long running task
}

**Reply by Munksgaard** [reply]
I'm in a similar situation as @grubeli, but I don't want to wait for the blocking computation to finish before returning a response. Is it possible to spawn a worker and then immediately return a response?

**Reply by taylor-shift** [reply]
@wezm Thanks for your reply to this question.

I am curious about the same as the above...I think we're (@Munksgaard and myself) are looking for something more along the lines of an async job queue where the task is queued up and a response sent immediately to the client, before processing begins on said task.

How would we do this in Rocket?

Apologies if I'm missing something obvious here, I am more used to working with Celery but understand that would be something different all-together.

**Reply by massimans** [reply]
> I'm in a similar situation as @grubeli, but I don't want to wait for the blocking computation to finish before returning a response. Is it possible to spawn a worker and then immediately return a response?

it is as simple as not not awaiting for it

```rust
use rocket::tokio;

#[get("/something")]
async fn your_handler() -> YourResponseType {
    let res = tokio::task::spawn_blocking(move || {
        // long running stuff in here
    });

    //  Generate response using result of long running task
}
```
